spec/Overview.html

```diff
@@ -1066,44 +1064,62 @@ <h3>CryptoKey interface members</h3>
           </dl>
         </section>
 
-        <section id="cryptokey-interface-clone">
-          <h3>Structured clone algorithm</h3>
+        <section id="cryptokey-interface-serializable">
+          <h3>Serialization and deserialization steps</h3>
           <p>
-            When a user agent is required to obtain a [= structured clone =]
-            of a {{CryptoKey}} object, it must run the following steps.
+            {{CryptoKey}} objects are <a>serializable objects</a>. Their [= serialization steps =],
+            given <var>value</var> and <var>serialized</var>, are:
           </p>
           <ol>
             <li>
-              Let <var>input</var> and <var>memory</var> be the corresponding inputs defined by the
-              [= structured clone | internal structured cloning algorithm =], where
-              <var>input</var> represents a {{CryptoKey}} object to be
-              cloned.
+              Set <var>serialized</var>.[[\Type]] to the <a>[[\type]]</a> internal slot of <var>value</var>.
             </li>
             <li>
-              Let <var>output</var> be a newly constructed {{CryptoKey}}
-              object.
+              Set <var>serialized</var>.[[\Extractable]] to the <a>[[\extractable]]</a> internal slot of <var>value</var>.
             </li>
             <li>
-              Let the <a>[[\type]]</a>, <a>[[\extractable]]</a>, <a>[[\algorithm]]</a>, and <a>[[\usages]]</a> internal slots of <var>output</var>
-              be set to the result of invoking the internal structured clone algorithm recursively
-              on the corresponding internal slots of <var>input</var>, with the slot contents as the
-              new "<var>input</var>" argument and <var>memory</var> as the new "<var>memory</var>"
-              argument.
+              Set <var>serialized</var>.[[\Algorithm]] to the [= sub-serialization =] of the
+              <a>[[\algorithm]]</a> internal slot of <var>value</var>.
             </li>
             <li>
-              Let the <a>[[\handle]]</a> internal slot of
-              <var>output</var> refer to the same cryptographic key data represented by the
-              <a>[[\handle]]</a> internal slot of <var>input</var>.
+              Set <var>serialized</var>.[[\Usages]] to the [= sub-serialization =] of the
+              <a>[[\usages]]</a> internal slot of <var>value</var>.
+            </li>
+            <li>
+              Set <var>serialized</var>.[[\Key]] to a serialized representation of the
```

Comment:
You cannot say serialized representation here and not say deserialized later on. Could it be sufficient to just set Key to handle? Or is this data mutable in some way? (Please excuse my lack of knowledge about what is going on.)

Reply:
The issue is that `[[handle]]` might refer to a key object in memory, for example (e.g., a pointer to an OpenSSL key object, or whatever underlying implementation is used) whereas this might have to go on disk. But maybe I can just say "Set <var>serialized</var>.[[Key]] to a representation of the cryptographic key data..." - I also don't like the double "serialized" here

Reply:
If I'm reading the specification correctly [[handle]] is a logical key in which case I think we can just assign and implementations can figure out how to make that work. Specifications don't operate on implementation-reality, they operate at a higher-level of abstraction. https://infra.spec.whatwg.org/#algorithm-conformance might help here.

What's a bit weird to me is that I cannot find the place that sets [[handle]]. It seems importKey() might be a way to mint a CryptoKey object, but it doesn't detail how [[handle]] is set.

Reply:
Right, fair enough. But even abstractly speaking, I think the name `[[handle]]` sort of implies that it's a "handle that refers to a cryptographic key stored elsewhere" - though the spec doesn't really say that explicitly, but it does say:

> The `CryptoKey` object represents an opaque reference to keying material that is managed by the user agent.

So storing that reference somewhere to me feels incorrect, as to me it seems to imply that they should serialize the handle / pointer, rather than the key material.

Re. initializing [[handle]], yeah I agree, it seems to be implied in the import key operations of each of the algorithms, with language such as:

> Let _key_ be a new `CryptoKey` associated with the relevant global object of `this` [HTML], and that represents the RSA public key identified by _publicKey_.

But it's fairly hand-wavy indeed. I think one (small) improvement could be to construct the `CryptoKey` seperately (perhaps earlier on in `importKey()`) and then replace this with "Set the [[handle]] internal slot of _key_ to a representation of the RSA public key identified by _publicKey_" or some such?

Reply:
If I had to guess perhaps the editors went with [[handle]] rather than [[key]] to make it clear the key is/can be stored in a different place not directly accessible from the process the `CryptoKey` object is in. In that case though it still seems to me like you should just set, but perhaps rather than calling it [[Key]] in the serialization you should call it [[Handle]].

Reply:
Hmhm. But if [[handle]] refers to a key in a different place / process, and we store that in IndexedDB, for example, and retrieve it again later, then that key will most likely be gone, no? And in this case, it's not so much that the implementation should "figure out how to keep it alive" I think, rather, to me it seems that the key material itself should be serialized, not the handle that refers to it.

Reply:
From a specification-perspective it won't be gone as there's nothing that said it could be gone. I suspect we reached the point where some input from @sleevi would help though. 😊

Reply:
Happy to help, although this is definitely an area where "Tried to do it right, but may have got it catastrophically wrong". This was one of the more challenging aspects of the spec, in part because of the lack of consensus to eliminate certain use cases.

The core of the issue was a lack of agreement about whether WebCrypto implementations MUST be software-backed, or whether they can support other systems. Similarly, of those that are software-backed, WebCrypto needed to play within the API constraints of major cryptographic libraries, since it was unclear **how** it would be implemented. This in particular had a lot of intersection for a use case some (e.g. Netflix) felt was very important, which was the ability to make guarantees about how the key object is protected from other applications on the device running the user agent.

So the intent was to try to find a way to specify some degree of persistence, without specifying that the raw key material itself needed to be provided here. For example, an implementation might use a TPM with an opaque (TPM-wrapped) blob as the persistence handle. However, for efficiency, "handle" might be a reference to an active TPM session (where the 'raw' key is loaded into the TPM's memory). If the TPM evicted that blob from memory, the implementation could reload it back into the TPM when another operation was performed on the `CryptoKey`. Or, on Windows or macOS, the key could be implemented via CNG/SecKey, where the "actual" key is stored on disk somewhere managed by the OS, but the calling application has an (OS) handle to it to perform operations.

To @annevk 's point, yes, this absolutely raised the scenario in which the key, managed on this external storage, goes away. The spec tries to ensure that a `CryptoKey` is still reconstructed, but it does mean that operations that attempt to use it will fail with `OperationError` if the handle to the OS-managed resource is no longer available.

By comparison, this would be similar to the Cache API implementation in which the underlying OS deleted a file from the backing store, but didn't update the index of what resources were in the Cache. When retrieving the object from the Cache API programmatically, it'd fail, because the underlying disk resource was now corrupted. I believe there's also some media APIs (related to WebGL or textures?) that have a property where the GPU could evict the texture, but the object is otherwise usable for as long as the media device has it.

To @annevk 's question about naming and such - yeah, the 'handwave' intent here was to capture that the internal-slot is an implementation-specific opaque maybe-key-maybe-not called `Handle`, and that an implementation must be able to refer to the same logical object in the future (e.g. when deserializing), even though that object may not exist. Again, the poor comparison here might be how a Cache API implementation may store individual resources as files with a particular file path, where the filepath is never exposed to the web. Further, the concept 'file path' doesn't make it into the API, because you **could** implement Cache API using, say, a SQLite database, where instead of file paths, you used row unique IDs in a table.

https://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-internal-slots tried to capture this, but perhaps poorly. If the concern is the phrasing 'serialized' here, then yes, I think we could drop it, and just say "set `[[handle]]` to handle", and let implementations figure out what precisely that means (e.g. storing the key name if using CNG, storing the underlying key material if using something like BoringSSL, a PKCS#11 URI if using NSS with external storage, etc)

Reply:
@sleevi Thanks for the background. Do you know if there are any implementations where the key material can indeed be "gone" after deserialization? E.g., does ChromeOS's implementation allow storing hardware-backed keys in IndexedDB? (I guess anyway in that case the hardware is built-in so probably it's not likely to disappear?)

Personally, I think that to properly support hardware-backed keys which can "go away", it would be nice to have a way for the web app to indicate that they would be able to deal with that (and perhaps have a separate case for those keys during serialization) and require all other keys to be persistent (as I suspect a lot of code already depends on this anyway), e.g. a `generateHardwareKey` which is mostly identical to `generateKey`, and still requires storing a reference in IndexedDB for persistence, but where the key is allowed to go away (though perhaps removable hardware is too niche of a use case for this). But yeah, I suppose for platforms that have built-in hardware, theoretically speaking they could already use that today when you call `generateKey` with `extractable=false`, so to support that use case we would indeed have to store the handle rather than the key material.

@annevk If we store the handle, do you think we should call the sub-serialization algorithm? On the one hand, I don't think it would really know how to serialize it, on the other hand, `serialized` is a [`Record`](https://tc39.es/ecma262/#sec-list-and-record-specification-type), and ECMA262 wants its values to either be an ECMAScript value or another `Record`. So maybe some handwavy "set _serialized_.[[Handle]] to a representation of [[handle]]" is still required?

Reply:
> @sleevi Thanks for the background. Do you know if there are any implementations where the key material can indeed be "gone" after deserialization? 

I believe Edge was/(is?) been such a case. They process isolated the key storage under the EdgeHTML implementation and relied on DPAPI for protection, except that DPAPI keys can be lost under certain circumstances (related to offline password changes), which would make decrypting the keys impossible.

> E.g., does ChromeOS's implementation allow storing hardware-backed keys in IndexedDB? (I guess anyway in that case the hardware is built-in so probably it's not likely to disappear?)

Most SEE hardware is extremely limited in storage, which is what my previous reply was referring to. The hardware itself doesn’t store all the keys - rather, it vends an opaque blob wrapped by a key that is in hardware, and then you feed that opaque blob in later. The opaque blob is basically “your” key wrapped by the hardware key, and it’s loaded into the hardware RAM.

This is part of TPM, and thus means it is possible for keys not to be unwrapped (e.g. a TPM reset).

I don’t believe ChromeOS fully implements structured clone for these keys (recall, it’s in an extension namespace that is ACL restricted, but which otherwise follows Web Crypto semantics), but I also believe that’s being fixed - in which case, keys can totally go away.

> Personally, I think that to properly support hardware-backed keys which can "go away", it would be nice to have a way for the web app to indicate that they would be able to deal with that (and perhaps have a separate case for those keys during serialization) and require all other keys to be persistent (as I suspect a lot of code already depends on this anyway).

Right, the problem with this is fundamental (and part of the past discussions): the web author has no way of actually knowing that the UA is doing this in a reliable/trustworthy way. That is, you can always polyfill on top of it to lie. The only way around that is protected attributes and attestation, which is just a fancy way of saying “DRM supercookies”.

> But yeah, I suppose for platforms that have built-in hardware, theoretically speaking they could already use that today when you call `generateKey` with `extractable=false`, so to support that use case we would indeed have to store the handle rather than the key material.

Even without hardware, it’s a separation of implementation concerns. The big issue was largely wanting to work with existing crypto APIs, because of the compliance concerns with a user agent shipping its own implementation. That’s why it was left opaque as much as possible.

Reply:
Right, I wasn't debating the value of having a handle in general, just whether, when serializing for storage, the implementation should export and serialize the key material or not (and I also wasn't trying to argue that we should force the UA to keep it in IDB forever, just that _if_ it's in IDB, and it's deserialized, that I imagine that there's a lot of code that assumes that it can be used).

But yeah, I also see the value in doing something like Edge is/was doing. I'll change the wording :+1:

Reply:
In HTML we don't follow the rule that Record field values have to be Records or JavaScript values and effectively store abstract values in them (such as "bitmap data"). This doesn't seem problematic to me, but it's probably something that needs to be cleaned up at some point. cc @domenic

It sounds to me what should happen is that we do say that we serialize [[handle]] in an abstract-manner. And then in the deserializaton steps we deserialize [[Key]] to a [[handle]] and say that if that fails for some reason, a "`DataCloneError`" `DOMException` is thrown. (At least, there's nothing in the specification that suggests `CryptoKey` can operate without a [[handle]] so that seems like what should happen.)

Reply:
I see. I guess in that case I wouldn't suggest that [[handle]] is serialized and deserialized here. I.e., back to set [[Handle]] to [[handle]]. It's just an immutable device to look up a key and when used it may fail to return a key.